projs/shadow/shadow-engine/core/inc/event-bus/event_bus.h

```diff
@@ -0,0 +1,139 @@
+#pragma once
+
+#include <string>
+#include <functional>
+#include <map>
+#include <memory>
+#include <any>
+
+#include "SHObject.h"
+#include "events.h"
+
+namespace ShadowEngine::EventBus {
+
+    template<class T, class U>
+    concept Derived = std::is_base_of<T, U>::value;
+
+    template<typename T>
+    concept SHObjectConcept = Derived<ShadowEngine::SHObject, T>;
+
+
+
+
+    template<typename T>
+    concept EventType = Derived<Event, T>;
+
+    template<EventType T>
+    struct Subscription {
+        using Callback = std::function<const void(T &)>;
+
+        Callback callback;
+
+        std::weak_ptr<ShadowEngine::SHObject> binding;
+
+        explicit Subscription(Callback call): callback(call){}
+    };
+
+    class EventDispatcherBase {
+    public:
+        virtual ~EventDispatcherBase() = default;
+
+        virtual void call(ShadowEngine::SHObject &obj) = 0;
+    };
+
+    template<EventType T>
+    class EventDispatcher : public EventDispatcherBase {
+    public:
+        using SubRef = std::shared_ptr<Subscription<T>>;
+
+    protected:
+        using SubscriptionList = std::vector<SubRef>;
+
+        SubscriptionList subscriptionList;
+
+    public:
+        std::shared_ptr<Subscription<T>> subscribe(const std::shared_ptr<ShadowEngine::SHObject> &binding, std::function<const void(T &)> func) {
+            auto sub = std::make_shared<Subscription<T>>(func);
+            subscriptionList.push_back(sub);
+            return sub;
+        }
+
+        void unsubscribe(const std::shared_ptr<Subscription<T>> ref) {
+            subscriptionList.erase(std::remove(subscriptionList.begin(), subscriptionList.end(), ref), subscriptionList.end());
+        }
+
+        void call(ShadowEngine::SHObject &obj) override {
+            this->call_impl((T&)obj);
+        }
+
+        void call_impl(T &event) {
+            for (int i = 0; i < subscriptionList.size(); ++i) {
+                subscriptionList[i]->callback(event);
+            }
+        }
+    };
+
+    template<EventType T>
+    class EventDispatcherThreaded : public EventDispatcher<T> {
+        std::mutex mObserversMutex;
+
+    public:
+
+        void call(ShadowEngine::SHObject &obj) override {
+            std::lock_guard<std::mutex> lLockGuard(mObserversMutex);
+
+            this->call_impl((T&)obj);
+        }
+    };
+
+    class EventBus {
+
+        template<class T,
+                class V =
+                std::conditional<
+                        std::is_base_of<Event,T>::value,
+                        EventDispatcher<T>,
+                        EventDispatcherThreaded<T>
+                >::type>
+        std::shared_ptr<V> getDispatcher(){
+            std::shared_ptr<V> dis;
+
+            if (dispatchers.contains(T::TypeId())) {
+                dis = std::reinterpret_pointer_cast<V>(dispatchers[T::TypeId()]);
+            } else {
+                dis = std::make_shared<V>();
```

Comment:
Wait, does this make a NEW DISPATCHER for each type of event that we try to register?
I seriously need some doc to confirm here, because that's going to have EXPONENTIAL runtime.

Reply:
Okay. Let me try to break this down.

The Source Engine has 380 events (Source + Source Server + a game implementation; TF2 for example). 
If we assume that half of those are fired once during setup, and half are fired at least a dozen times a second (which averages out for per-frame events and gameplay events such as Control Point captures) then we'll be running this function 136,800 times per second.

Each time we call it, we have to call through the Type ID system and index the dispatchers list, which may take multiple calls due to a deep event hierarchy.

If we take the deepest event hierarchy from Source, which i believe is teamplay_win_panel, the Derived template will have to go through the event up to 8 times...

which means that on average, this single function will have a runtime complexity of at least O(n^k), due to the split nature of the dispatchers and the templates that manage the event bus and dispatchers.

To put it in perspective, that's the equivalent of 1.22E41 calls, assuming that every event is as "deep" and expensive as teamplay_win_panel.

We need to change this. 
Urgently.

Reply:
Oh, and that's not even including the efficiency of std::map operator[] when searching for an arbitrary uint64_t.